TheAmazingAudioEngine/AEAudioUnitFileStreamer.m

```diff
@@ -0,0 +1,395 @@
+//
+//  Permission is granted to anyone to use this software for any purpose,
+//  including commercial applications, and to alter it and redistribute it
+//  freely, subject to the following restrictions:
+//
+//  1. The origin of this software must not be misrepresented; you must not
+//     claim that you wrote the original software. If you use this software
+//     in a product, an acknowledgment in the product documentation would be
+//     appreciated but is not required.
+//
+//  2. Altered source versions must be plainly marked as such, and must not be
+//     misrepresented as being the original software.
+//
+//  3. This notice may not be removed or altered from any source distribution.
+//
+
+#import "AEAudioUnitFileStreamer.h"
+#import <libkern/OSAtomic.h>
+
+@implementation AEAudioUnitFileStreamer
+
+@dynamic duration,currentTime, playbackDelayInSeconds, channelIsPlaying;
+
+// MARK:  Debug Properties
+@synthesize currentFrame = _playhead, totalFrames = _lengthInFrames;
+
+// Notifications
+static void notifyPlaybackStopped(AEAudioController *audioController, void *userInfo, int length) {
+    AEAudioUnitFileStreamer *THIS = (__bridge AEAudioUnitFileStreamer*)*(void**)userInfo;
+    
+    if(OSAtomicCompareAndSwap32(YES,NO,&THIS->_playbackStoppedCallbackScheduled)) {
+        THIS.channelIsPlaying = NO;
+        THIS->_playhead = 0;
+
+        if ( THIS.completionBlock )
+            THIS.completionBlock();
+        
+        NSError* error;
+        [THIS setupPlayRegion:&error];
+        if(error)
+            NSLog(@"Error on playback stopped callback: %@",error.description);
+    }
+}
+
+// MARK: Setup and Teardown functions
+- (BOOL)setupPlayRegion: (NSError**) error
+{
+    OSStatus result = -1;
+    
+    if(_audioUnitFile)
+    {
+        // Set playhead to starting point if it is beyond the end of the song.
+        if(_playhead >= _lengthInFrames)
+            _playhead = 0;
+        
+        // Reset the audio unit before modifying properties
+        AudioUnit au = AEAudioUnitChannelGetAudioUnit(self);
+        checkResultForError(AudioUnitReset(au, kAudioUnitScope_Global, 0),"AudioUnitReset()","Error resetting audio unit",NO,result,error)
+        
+        // Create the Scheduled Audio File Region struct
+        ScheduledAudioFileRegion region;
+        memset (&region.mTimeStamp, 0, sizeof(region.mTimeStamp));
+        
+        // Need to manually validate the time stamp and set it's start time to 0
+        region.mTimeStamp.mFlags = kAudioTimeStampSampleTimeValid;
+        region.mTimeStamp.mSampleTime = _playbackDelay;
+        
+        // No process will be called on completion
+        region.mCompletionProc = NULL;
+        region.mCompletionProcUserData = NULL;
+        
+        // Point to the audio unit file
+        region.mAudioFile = _audioUnitFile;
+        region.mLoopCount = 0;
+        region.mStartFrame = _playhead;
+        region.mFramesToPlay = _lengthInFrames;// - _playhead;
```

Comment:
Line 94 is the entire mechanism of how the player deals with seeking, so please don't comment it out! ;)

When _playhead is 0, the file plays at the beginning. If _playhead > 0 then the file will play from the _playhead frame. If _playhead < 0, then the audio unit stops working (no audio comes out), but there are no crashes. So the only change I might make is adding

if(_playhead < 0) region.mStartFrame = 0;